src/cockpit/channel.py

```diff
@@ -275,7 +277,20 @@ def send_data(self, data: bytes) -> bool:
         should stop writing for now.  In that case, `.do_resume_send()` will be
         called later when there is more room.
         """
-        self.send_channel_data(self.channel, data)
+        if self.is_binary:
+            self.send_channel_data(self.channel, data)
+        else:
+            # for text channels we must avoid splitting UTF-8 multi-byte characters,
+            # as these can't be sent to a WebSocket (and are often confusing for text streams as well)
+            if self.decoder is None:
+                self.decoder = codecs.getincrementaldecoder('utf-8')(errors='strict')
+            try:
+                block = self.decoder.decode(data)
+            except UnicodeDecodeError as exc:
+                raise ChannelError('protocol-error', message=str(exc)) from exc
+            self.send_channel_data(self.channel, block.encode())
+            # note: It is possible that after the last send_data() call there is a residual state in the decoder.
+            # That can only happen if a text channel receives invalid/truncated UTF-8 data. So we just don't send it.
```

Comment:
Let's do something about this: If there's residual state in `self.close()` it could at least log a warning. However, this should only happen if the channel was closed without a `problem` code, as errors can happen at any time. @allisonkarlitskaya WDYT?

Reply:
Thanks, I wasn't sure about that -- but right, this is the reading side -- we'd always get message-message-message-done.

Reply:
I did that now.